app/controllers/shared/api.php

```diff
@@ -19,234 +19,267 @@
+        $usage
+            ->setParam('projectId', $project->getId())
+            ->setParam('httpRequest', 1)
+            ->setParam('httpUrl', $request->getHostname() . $request->getURI())
+            ->setParam('httpMethod', $request->getMethod())
+            ->setParam('httpPath', $route->getPath())
+            ->setParam('networkRequestSize', 0)
+            ->setParam('networkResponseSize', 0)
+            ->setParam('storage', 0)
+        ;
+
+        $deletes->setProject($project);
+        $database->setProject($project);
+    });
+
+App::init()
+    ->groups(['auth'])
+    ->inject('utopia')
+    ->inject('request')
+    ->inject('project')
+    ->action(function (App $utopia, Request $request, Document $project) {
+
+        $route = $utopia->match($request);
+
+        $isPrivilegedUser = Auth::isPrivilegedUser(Authorization::getRoles());
+        $isAppUser = Auth::isAppUser(Authorization::getRoles());
+
+        if ($isAppUser || $isPrivilegedUser) { // Skip limits for app and console devs
+            return;
+        }
 
-    if (!empty($events->getEvent())) {
-        if (empty($events->getPayload())) {
-            $events->setPayload($response->getPayload());
+        $auths = $project->getAttribute('auths', []);
+        switch ($route->getLabel('auth.type', '')) {
+            case 'emailPassword':
+                if (($auths['emailPassword'] ?? true) === false) {
+                    throw new Exception('Email / Password authentication is disabled for this project', 501, Exception::USER_AUTH_METHOD_UNSUPPORTED);
+                }
+                break;
+
+            case 'magic-url':
+                if ($project->getAttribute('usersAuthMagicURL', true) === false) {
+                    throw new Exception('Magic URL authentication is disabled for this project', 501, Exception::USER_AUTH_METHOD_UNSUPPORTED);
+                }
+                break;
+
+            case 'anonymous':
+                if (($auths['anonymous'] ?? true) === false) {
+                    throw new Exception('Anonymous authentication is disabled for this project', 501, Exception::USER_AUTH_METHOD_UNSUPPORTED);
+                }
+                break;
+
+            case 'invites':
+                if (($auths['invites'] ?? true) === false) {
+                    throw new Exception('Invites authentication is disabled for this project', 501, Exception::USER_AUTH_METHOD_UNSUPPORTED);
+                }
+                break;
+
+            case 'jwt':
+                if (($auths['JWT'] ?? true) === false) {
+                    throw new Exception('JWT authentication is disabled for this project', 501, Exception::USER_AUTH_METHOD_UNSUPPORTED);
+                }
+                break;
+
+            default:
+                throw new Exception('Unsupported authentication route', 501, Exception::USER_AUTH_METHOD_UNSUPPORTED);
+                break;
+        }
+    });
+
+App::shutdown()
+    ->groups(['api'])
```

Comment:
The rest of PR puts groups under injection. Let's do it here too for consistency.